File: backend/omninest-file/src/main/java/com/omninest/modules/file/dto/FileUploadSessionDto.java
package com.omninest.modules.file.dto;

import io.swagger.v3.oas.annotations.media.Schema;
import java.time.Instant;
import java.util.List;
import java.util.UUID;

@Schema(description = "文件上传会话信息")
public record FileUploadSessionDto(
        @Schema(description = "会话 ID") UUID id,
        @Schema(description = "父文件夹 ID") UUID parentId,
        @Schema(description = "上传 ID", example = "upload-abc123") String uploadId,
        @Schema(description = "文件名", example = "文档.pdf") String fileName,
        @Schema(description = "文件大小（字节）", example = "1048576") long sizeBytes,
        @Schema(description = "分片大小（字节）", example = "5242880") int partSizeBytes,
        @Schema(description = "总分片数", example = "3") int totalParts,
        @Schema(description = "MIME 类型", example = "application/pdf") String mimeType,
        @Schema(description = "会话状态", example = "PENDING") String status,
        @Schema(description = "存储桶名称") String bucket,
        @Schema(description = "对象键") String objectKey,
        @Schema(description = "上传 URL") String uploadUrl,
        @Schema(description = "分片信息列表") List<FileUploadPartDto> parts,
        @Schema(description = "会话过期时间") Instant expiresAt
) {
}

File: frontend/lib/features/files/data/dtos/file_upload_session_dto.dart
import 'package:omninest/features/files/domain/file_upload_session.dart';

class FileUploadPartDto {
  const FileUploadPartDto({
    required this.partNumber,
    required this.sizeBytes,
    required this.status,
    required this.uploadUrl,
    this.eTag,
  });

  factory FileUploadPartDto.fromJson(Map<String, dynamic> json) {
    return FileUploadPartDto(
      partNumber: _asInt(json['partNumber']),
      sizeBytes: _asInt(json['sizeBytes']),
      status: json['status']?.toString() ?? 'PENDING',
      eTag: json['eTag']?.toString(),
      uploadUrl: json['uploadUrl']?.toString(),
    );
  }

  final int partNumber;
  final int sizeBytes;
  final String status;
  final String? eTag;
  final String? uploadUrl;

  FileUploadPart toDomain() {
    return FileUploadPart(
      partNumber: partNumber,
      sizeBytes: sizeBytes,
      status: status,
      eTag: eTag,
      uploadUrl: uploadUrl,
    );
  }
}

class FileUploadSessionDto {
  const FileUploadSessionDto({
    required this.id,
    required this.uploadId,
    required this.parentId,
    required this.fileName,
    required this.sizeBytes,
    required this.partSizeBytes,
    required this.totalParts,
    required this.mimeType,
    required this.status,
    required this.bucket,
    required this.objectKey,
    required this.uploadUrl,
    required this.parts,
    required this.expiresAt,
  });

  factory FileUploadSessionDto.fromJson(Map<String, dynamic> json) {
    final rawParts = json['parts'];
    return FileUploadSessionDto(
      id: json['id'].toString(),
      uploadId: json['uploadId']?.toString() ?? '',
      parentId: json['parentId']?.toString(),
      fileName: json['fileName']?.toString() ?? '',
      sizeBytes: _asInt(json['sizeBytes']),
      partSizeBytes: _asInt(json['partSizeBytes']),
      totalParts: _asInt(json['totalParts']),
      mimeType: json['mimeType']?.toString() ?? 'application/octet-stream',
      status: json['status']?.toString() ?? 'CREATED',
      bucket: json['bucket']?.toString() ?? '',
      objectKey: json['objectKey']?.toString() ?? '',
      uploadUrl: json['uploadUrl']?.toString(),
      parts:
          rawParts is List
              ? rawParts
                  .whereType<Map<String, dynamic>>()
                  .map(FileUploadPartDto.fromJson)
                  .map((dto) => dto.toDomain())
                  .toList()
              : const [],
      expiresAt:
          DateTime.tryParse(json['expiresAt']?.toString() ?? '')?.toLocal(),
    );
  }

  final String id;
  final String uploadId;
  final String? parentId;
  final String fileName;
  final int sizeBytes;
  final int partSizeBytes;
  final int totalParts;
  final String mimeType;
  final String status;
  final String bucket;
  final String objectKey;
  final String? uploadUrl;
  final List<FileUploadPart> parts;
  final DateTime? expiresAt;

  FileUploadSession toDomain() {
    return FileUploadSession(
      id: id,
      uploadId: uploadId,
      parentId: parentId,
      fileName: fileName,
      sizeBytes: sizeBytes,
      partSizeBytes: partSizeBytes,
      totalParts: totalParts,
      mimeType: mimeType,
      status: status,
      bucket: bucket,
      objectKey: objectKey,
      uploadUrl: uploadUrl,
      parts: parts,
      expiresAt: expiresAt,
    );
  }
}

int _asInt(Object? value) {
  return switch (value) {
    final int number => number,
    final num number => number.toInt(),
    final String text => int.tryParse(text) ?? 0,
    _ => 0,
  };
}
